File: cpp/3.0/rdt.cc

#include "rdt.hh"
#include "timer.h" // timer in secs

#include <iostream>
#include <unistd.h>
#include <cstring>
#include <arpa/inet.h>
#include <random>
#include <fcntl.h> // IO control
#include <time.h>  // sleep for ms

using namespace std;

Sender::Sender(const Address addr, const Address paddr) : addr(addr), paddr(paddr)
{
    this->seq = 0;
    this->sockfd = socket(AF_INET, SOCK_DGRAM, 0);
    if (this->sockfd == -1)
    {
        cerr << "Failed to create UDP socket." << std::endl;
        return;
    }

    ///// set unblock mode /////
    // int flags = fcntl(this->sockfd, F_GETFL, 0);
    // fcntl(sockfd, F_SETFL, flags | O_NONBLOCK);
    timeval timeout;
    timeout.tv_sec = 0;
    timeout.tv_usec = RECV_TIMEOUT;
    setsockopt(this->sockfd, SOL_SOCKET, SO_RCVTIMEO, (void *)&timeout, sizeof(timeout));
    ///// set unblock mode /////

    memset(&(this->sock_addr), 0, sizeof(this->sock_addr));

    sockaddr_in *sin_addr_p = (sockaddr_in *)&(this->sock_addr); // intenet socket pointer
    sin_addr_p->sin_family = AF_INET;                                 // 设置为IP通信
    sin_addr_p->sin_addr.s_addr = inet_addr(addr.ip.data());          // 设置IP地址
    sin_addr_p->sin_port = htons(addr.port);                          // 设置端口号

    memset(&this->sock_paddr, 0, sizeof(this->sock_paddr));

    sockaddr_in *sin_paddr_p = (sockaddr_in *)&(this->sock_paddr); // intenet socket pointer
    sin_paddr_p->sin_family = AF_INET;                                  // 设置为IP通信
    sin_paddr_p->sin_addr.s_addr = inet_addr(paddr.ip.data());          // 设置IP地址
    sin_paddr_p->sin_port = htons(paddr.port);                          // 设置端口号

    if (bind(this->sockfd, &(this->sock_addr), sizeof(this->sock_addr)) < 0)
    {
        cerr << "Failed to bind UDP socket." << std::endl;
        return;
    }
}

Sender::~Sender()
{
    close(this->sockfd);
}

bool Sender::rdt_send(const uint8_t *data, const int dsize)
{
    cout << "this->seq: " << this->seq << endl;

    make_pkt_beta(DATA_FLAG,
                  this->seq,
                  dsize,
                  data,
                  this->send_buffer,
                  BUFFER_SIZE);

    Pocket *send_pkt_p = (Pocket *)(this->send_buffer);

    cout << "+ SEND" << endl;
    print_pkt(send_pkt_p);

    udt_send(send_pkt_p);

    /////// start timer //////////
    // time_t start_time = time(NULL);
    double start_time, curr_time;
    GET_TIME(start_time);
    /////// start timer //////////

    while (true)
    {
        GET_TIME(curr_time);
        if (curr_time - start_time > TIMEOUT)
        {
            cout << "Time out: resend pkt_" << send_pkt_p->seq << endl;
            udt_send(send_pkt_p);
            GET_TIME(start_time);
        }

        if (udt_recv(this->recv_buffer) == false)
        {
            continue;
        }

        Pocket *recv_pkt_p = (Pocket *)(this->recv_buffer);
        cout << "- RECV" << endl;
        print_pkt(recv_pkt_p);

        if (is_corrupt(recv_pkt_p)) // corrupt
        {
            cout << "1 corrupt pkt!" << endl;
            udt_send(send_pkt_p);
            continue;
        }
        // not corrupt
        if (recv_pkt_p->flag != ACK_FLAG) // not ACK pkt
        {
            cout << "2 recv pkt is not ACK!" << endl;
            udt_send(send_pkt_p);
            continue;
        }
        // ACK pkt
        if (recv_pkt_p->seq != this->seq) // wrong seq
        {
            cout << "3 wrong ack seq: recv(" << recv_pkt_p->seq << "), await(" << this->seq << ")" << endl;
            udt_send(send_pkt_p);
            continue;
        }
        // right pkt
        cout << "!! right ack seq: recv(" << recv_pkt_p->seq << "), await(" << this->seq << ")" << endl;

        cout << "this->seq: " << this->seq << endl;
        cout << "NEXT(this->seq): " << NEXT(this->seq) << endl;
        this->seq = NEXT(this->seq);

        break;
    }

    return true;
}

Receiver::Receiver(const Address addr, const Address paddr) : addr(addr), paddr(paddr)
{
    this->seq = 0;
    this->sockfd = socket(AF_INET, SOCK_DGRAM, 0);
    if (this->sockfd == -1)
    {
        cerr << "Failed to create UDP socket." << std::endl;
        return;
    }
    memset(&(this->sock_addr), 0, sizeof(this->sock_addr));

    sockaddr_in *sin_addr_p = (struct sockaddr_in *)&(this->sock_addr); // intenet socket pointer
    sin_addr_p->sin_family = AF_INET;                                 // 设置为IP通信
    sin_addr_p->sin_addr.s_addr = inet_addr(addr.ip.data());          // 设置IP地址
    sin_addr_p->sin_port = htons(addr.port);                          // 设置端口号

    memset(&this->sock_paddr, 0, sizeof(this->sock_paddr));

    sockaddr_in *sin_paddr_p = (struct sockaddr_in *)&(this->sock_paddr); // intenet socket pointer
    sin_paddr_p->sin_family = AF_INET;                                  // 设置为IP通信
    sin_paddr_p->sin_addr.s_addr = inet_addr(paddr.ip.data());          // 设置IP地址
    sin_paddr_p->sin_port = htons(paddr.port);                          // 设置端口号

    if (bind(this->sockfd, &(this->sock_addr), sizeof(this->sock_addr)) < 0)
    {
        cerr << "Failed to bind UDP socket." << std::endl;
        return;
    }
}

Receiver::~Receiver()
{
    close(this->sockfd);
}

bool Receiver::rdt_recv(uint8_t *data_buffer, const int db_size)
{

    while (true)
    {
        if (udt_recv(this->recv_buffer) == false)
        {
            continue;
        }
        Pocket *recv_pkt_p = (Pocket *)(this->recv_buffer);
        cout << "- RECV" << endl;
        print_pkt(recv_pkt_p);
        ///////////////////
        // random_bits_err(recv_pkt_p);
        // sleep(1);
        ///////////////////
        if (is_corrupt(recv_pkt_p)) // corrupt
        {
            cout << "1 corrupt pkt! send: ACK_" << NEXT(this->seq) << ", need pkt_" << this->seq << endl;

            make_pkt_beta(ACK_FLAG, NEXT(this->seq), 0, 0, this->send_buffer, BUFFER_SIZE);
            Pocket *send_pkt_p = (Pocket *)(this->send_buffer);

            udt_send(send_pkt_p);

            cout << "+ SEND" << endl;
            print_pkt(send_pkt_p);
            continue;
        }
        // not corrupt
        if (recv_pkt_p->flag != DATA_FLAG) // not data pkt
        {
            cout << "2 recv pkt is not data pkt!" << endl;

            make_pkt_beta(ACK_FLAG, NEXT(this->seq), 0, 0, this->send_buffer, BUFFER_SIZE);
            Pocket *send_pkt_p = (Pocket *)(this->send_buffer);
            udt_send(send_pkt_p);

            cout << "+ SEND" << endl;
            print_pkt(send_pkt_p);
            continue;
        }
        // data pkt
        if (recv_pkt_p->seq != this->seq) // wrong seq
        {
            cout << "3 wrong data seq: recv(" << recv_pkt_p->seq << "), await(" << this->seq << "), send(" << NEXT(this->seq) << ")" << endl;

            make_pkt_beta(ACK_FLAG, NEXT(this->seq), 0, 0, this->send_buffer, BUFFER_SIZE);
            Pocket *send_pkt_p = (Pocket *)(this->send_buffer);
            udt_send(send_pkt_p);

            cout << "+ SEND" << endl;
            print_pkt(send_pkt_p);
            continue;
        }
        // right seq pkt
        if (recv_pkt_p->size > db_size)
        {
            cerr << "data size > db_size!" << endl;
            break;
        }

        /// sleep to trigger timeout ///
        // sleep(1); // 1s
        timespec ts;
        ts.tv_sec = 0;
        ts.tv_nsec = RECV_SLEEP;
        nanosleep(&ts, NULL);
        /// sleep to trigger timeout ///

        cout << "!! right data pkt: recv(" << recv_pkt_p->seq << "), await(" << this->seq << ")" << endl;

        make_pkt_beta(ACK_FLAG, this->seq, 0, 0, this->send_buffer, BUFFER_SIZE);
        Pocket *send_pkt_p = (Pocket *)(this->send_buffer);

        udt_send(send_pkt_p);

        cout << "+ SEND" << endl;
        print_pkt(send_pkt_p);

        cout << "this->seq: " << this->seq << endl;
        cout << "NEXT(this->seq): " << NEXT(this->seq) << endl;
        this->seq = NEXT(this->seq);

        // copy data to data_buffer
        memcpy(data_buffer, recv_pkt_p->data, recv_pkt_p->size);
        data_buffer[recv_pkt_p->size] = 0; // string end '\0'

        break;
    }

    return true;
}

void Sender::udt_send(const Pocket *send_pkt_p)
{
    if (sendto(this->sockfd,
               (void *)send_pkt_p,
               sizeof(Pocket),
               0,
               &(this->sock_paddr),
               sizeof(this->sock_paddr)) < 0)
    {
        cerr << "Error sending packet" << std::endl;
    }
}

bool Sender::udt_recv(const uint8_t *recv_buffer)
{
    socklen_t paddr_len = sizeof(this->sock_paddr);

    int len = recvfrom(this->sockfd,
                       &(this->recv_buffer),
                       BUFFER_SIZE,
                       0,
                       &(this->sock_paddr),
                       &paddr_len);
    if (len == -1)
    {
        if (errno == EAGAIN)
        {
            // cout << "? recv time out!" << endl;
            return false;
        }
        cerr << "Failed to recv msg." << std::endl;
        return false;
    }
    return true;
}

void Receiver::udt_send(const Pocket *send_pkt_p)
{
    if (sendto(this->sockfd,
               (void *)send_pkt_p,
               sizeof(Pocket),
               0,
               &(this->sock_paddr),
               sizeof(this->sock_paddr)) < 0)
    {
        cerr << "Error sending packet" << std::endl;
    }
}
bool Receiver::udt_recv(const uint8_t *recv_buffer)
{
    {
        socklen_t paddr_len = sizeof(this->sock_paddr);

        int len = recvfrom(this->sockfd,
                           &(this->recv_buffer),
                           BUFFER_SIZE,
                           0,
                           &(this->sock_paddr),
                           &paddr_len);
        if (len == -1)
        {
            if (errno == EAGAIN)
            {
                // cout << "? recv time out!" << endl;
                return false;
            }
            cerr << "Failed to recv msg." << std::endl;
            return false;
        }
        return true;
    }
}

///// Helper Functions /////

bool make_pkt_beta(
    const uint32_t flag,
    const uint32_t seq,
    const uint32_t size,
    const uint8_t *data,
    const uint8_t *buffer,
    const int buffer_size)
{
    if ((size > DATA_MAX_SIZE) || (buffer_size < sizeof(Pocket)))
    {
        // data too big || buffer too small
        return false;
    }
    memset((void *)buffer, 0, buffer_size);

    Pocket *p = (Pocket *)buffer;

    p->flag = flag;
    p->seq = seq;
    p->size = size;
    memcpy(p->data, data, size);

    p->checksum = cal_checksum(p);
    return true;
}

void print_pkt(Pocket *p)
{
    if (is_corrupt(p))
    {
        cout << "pkt corrupt!" << endl;
    }
    cout << "== pkt checksum: " << p->checksum << endl;
    cout << "== pkt seq: " << p->seq << endl;
    cout << "== pkt size: " << p->size << endl;
}

uint32_t cal_checksum(const Pocket *p)
{
    uint32_t checksum = 0;
    uint8_t *pp = (uint8_t *)p;
    for (int i = CHECKSUM_START; i < sizeof(Pocket); i++)
    {
        checksum += pp[i];
    }
    return checksum;
}

bool is_corrupt(const Pocket *p)
{
    if (cal_checksum(p) == p->checksum)
    {
        return false;
    }
    return true;
}

#define RAND_THRES 9

void random_bits_err(Pocket *&p)
{
    random_device r;
    default_random_engine gen(r());
    uniform_int_distribution<int> udist(0, 10);
    int randint = udist(gen);
    // int randint = rand() % 10 + 1;
    // cout << "randint: " << randint << endl;
    if (RAND_THRES < randint)
    {
        return;
    }
    uint8_t *pp = (uint8_t *)p;
    pp[1] = 7 - pp[1];
    return;
}

File: python/3.0/rdt.py
"""
基于 udp 的可靠数据传输协议

rdt 3.0 添加:
    - 计时器

数据类型:
    - data 字节序列
    - pkt 为打包 {data, checksum, seq} 后的包字节序列
    - p 为字节序列解包后得到的对象
"""
import socket
import zlib
import pickle
import random
import time
RATE = 0  # bit err rate
TIME = 0.3
RECV_TIME = 0.1
RECVIVER_DELAY = 0.35
# helper functions


def cal_checksum(data):
    crc32 = zlib.crc32(data)
    return crc32


def random_bits_err(pkt):
    if random.random() > RATE:
        return pkt
    else:
        p = pickle.loads(pkt)
        p['data'] = "9999".encode()
        return pickle.dumps(p)


def is_corrupt(pkt):
    try:
        p = pickle.loads(pkt)
    except Exception as e:
        print("corrupted 0! {e}")
        return True
    if p['checksum'] != cal_checksum(p['data']):
        print("corrupted 1!")
        return True
    else:
        return False


def make_pkt(data, seq):
    checksum = cal_checksum(data)
    p = {
        'checksum': checksum,
        'seq': seq,
        'data': data
    }
    return pickle.dumps(p)


def udt_send(pkt, addr, paddr):
    """
    不可靠数据传输, send pkt from addr to paddr
    """
    with socket.socket(socket.AF_INET, socket.SOCK_DGRAM) as sock:
        sock.bind(addr)
        sock.sendto(pkt, paddr)


def udt_recv(addr):
    """
    不可靠数据接收, 监听发给addr的消息
    """

    with socket.socket(socket.AF_INET, socket.SOCK_DGRAM) as sock:
        try:
            sock.settimeout(RECV_TIME)
            sock.bind(addr)
            pkt, paddr = sock.recvfrom(1024)
            return pkt
        except socket.timeout:
            # print("监听超时")
            return None


class Sender:
    def __init__(self, addr, paddr):
        """
        recv_addr: (hostname, port) or (ipaddr, port)
        """
        self.addr = addr
        self.paddr = paddr
        self.winsz = 1
        self.seq = 0

    def rdt_send(self, data):
        print(f"== send pkt seq: {self.seq}")
        send_pkt = make_pkt(data, self.seq)
        udt_send(send_pkt, self.addr, self.paddr)
        # start timer
        start_time = time.time()
        while True:
            # check timer
            if time.time() - start_time > TIME:  # 超时 resend, start timer
                print("time out: resend!")
                udt_send(send_pkt, self.addr, self.paddr)
                start_time = time.time()
                continue

            recv_pkt = udt_recv(self.addr)
            if recv_pkt is None:
                continue
            isco = is_corrupt(recv_pkt)
            # 坏包 -> continue, 不转移
            if isco:
                print(f"坏包: resend: {self.seq}, continue")
                continue

            recv_p = pickle.loads(recv_pkt)
            # 错序包 -> continue, 不转移
            if recv_p['seq'] == self.last_seq:
                print(
                    f"错序包: recv seq: {recv_p['seq']}, resend: {self.seq}, continue")
                continue
            # 好包 and 正序包 -> 进入下一个等待状态
            elif (not isco) and (recv_p['seq'] == self.seq):
                print(
                    f"收到好包: recv seq: {recv_p['seq']}, sended: {self.seq}, dt: {time.time() - start_time}")

                break
            else:
                raise Exception("Sender else!")
        # 更新seq
        self.seq = (self.seq + 1) % (self.winsz + 1)
        # stop timer

    @property
    def last_seq(self):
        return (self.seq - 1) % (self.winsz + 1)


class Receiver:
    def __init__(self, addr, paddr):
        self.addr = addr
        self.paddr = paddr
        self.winsz = 1
        self.seq = 0

    def rdt_recv(self):
        while True:
            recv_pkt = udt_recv(self.addr)
            if recv_pkt is None:
                continue
            recv_pkt = random_bits_err(recv_pkt)  # bit 差错包
            isco = is_corrupt(recv_pkt)
            # 坏包 -> 重传, 不转移
            if isco:
                print(
                    f"坏包: send: ACK_{self.last_seq}, need: resend {self.seq}")
                send_pkt = make_pkt("ACK".encode(), self.last_seq)
                udt_send(send_pkt, self.addr, self.paddr)
                continue

            recv_p = pickle.loads(recv_pkt)
            # 错序包
            if (recv_p['seq'] == self.last_seq):
                print(
                    f"错序包: recv seq:{recv_p['seq']}, send: ACK_{self.last_seq}")
                send_pkt = make_pkt("ACK".encode(), self.last_seq)
                udt_send(send_pkt, self.addr, self.paddr)
                continue
            # 好包：
            elif (not isco) and (recv_p['seq'] == self.seq):
                print(f"好包: recv seq: {recv_p['seq']}, send: ACK_{self.seq}")
                data = recv_p['data']
                send_pkt = make_pkt("ACK".encode(), self.seq)
                time.sleep(RECVIVER_DELAY)
                udt_send(send_pkt, self.addr, self.paddr)
                break
            else:
                raise Exception("Receicer else!")
        self.seq = (self.seq + 1) % (self.winsz + 1)
        return data

    @property
    def last_seq(self):
        return (self.seq - 1) % (self.winsz + 1)
